build-logic/src/main/kotlin/dokkabuild/tasks/MvnExec.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package dokkabuild.tasks
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.file.FileSystemOperations
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.*
+import org.gradle.api.tasks.Optional
+import org.gradle.api.tasks.PathSensitivity.NONE
+import org.gradle.api.tasks.PathSensitivity.RELATIVE
+import org.gradle.process.ExecOperations
+import org.gradle.work.NormalizeLineEndings
+import java.util.*
+import javax.inject.Inject
+
+/**
+ * Runs a Maven task.
+ *
+ * See `dokkabuild.setup-maven-cli.gradle.kts` for details on the Maven CLI installation.
+ */
+@CacheableTask
+abstract class MvnExec
+@Inject
+constructor(
+    private val exec: ExecOperations,
+    private val fs: FileSystemOperations,
+) : DefaultTask() {
+
+    /**
+     * Work directory.
+     *
+     * Be aware that any existing content will be replaced by [inputFiles].
+     */
+    @get:OutputDirectory
+    abstract val workDirectory: DirectoryProperty
+
+    /** Input files - will be synced to [workDirectory]. */
+    @get:InputFiles
+    @get:PathSensitive(RELATIVE)
+    @get:NormalizeLineEndings
+    abstract val inputFiles: ConfigurableFileCollection
+
+    @get:InputFile
+    @get:PathSensitive(NONE)
+    abstract val mvnCli: RegularFileProperty
+
+    @get:Input
+    abstract val arguments: ListProperty<String>
+
+    /** `-e` - Produce execution error messages. */
+    @get:Input
+    @get:Optional
+    abstract val showErrors: Property<Boolean>
+
+    /** `-B` - Run in non-interactive (batch) mode. */
+    @get:Input
+    @get:Optional
+    abstract val batchMode: Property<Boolean>
+
+    @TaskAction
+    fun exec() {
+        fs.sync {
+            from(inputFiles)
+            into(workDirectory)
+        }
+
+        val arguments = buildList {
+            addAll(arguments.get())
+            if (showErrors.orNull == true) add("--errors")
+            if (batchMode.orNull == true) add("--batch-mode")
+        }
+
+        exec.exec {
+            workingDir(workDirectory)
+            executable(mvnCli.get())
+            args(arguments)
+        }
+
+        makePropertiesFilesReproducible()
+    }
+
+    /**
+     * Remove non-reproducible timestamps from any generated [Properties] files.
+     */
+    private fun makePropertiesFilesReproducible() {
```

Comment:
Yes that's right, it will affect up-to-date checks.

If this task re-runs and produces the same output, except for the timestamp, then downstream tasks that use this output will have to re-run, even though the timestamp has no practical impact.

It will also affect Build Cache.

Reply:
Gradle has a WriteProperties task that is similar, and uses a util that removes the timestamp

https://github.com/gradle/gradle/blob/e9d7e47044e08be2a88a64cb75742a85136da0ff/platforms/core-runtime/base-services/src/main/java/org/gradle/internal/util/PropertiesUtils.java#L68-L81